src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1013 @@
+#pragma once
+
+#include <memory>
+#include <stack>
+#include <string>
+#include <vector>
+
+#include <cereal/cereal.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/types/value.hpp>
+
+namespace cereal {
+
+/**
+ * A base class that holds a shared_ptr to the binary data for a BSON document. If a class you are
+ * serializing contains any of the bsoncxx view types (b_utf8, b_document, b_array, b_binary), you
+ * must have that class inherit this base so that the views don't point to deallocated memory.
+ *
+ * When using the BSONInputArchive to deserialize a class that inherits from this base, the data
+ * will be automatically loaded into this class using setUnderlyingBSONData().
+ */
+class UnderlyingBSONDataBase {
+   public:
+    void setUnderlyingBSONData(std::shared_ptr<uint8_t> ptr, size_t size) {
+        _underlyingBSONData = ptr;
+        _underlyingBSONDataSize = size;
+    }
+
+    const bsoncxx::document::view getUnderlyingBSONData() {
+        return bsoncxx::document::view{_underlyingBSONData.get(), _underlyingBSONDataSize};
+    }
+
+   private:
+    std::shared_ptr<uint8_t> _underlyingBSONData;
+    size_t _underlyingBSONDataSize;
```

Comment:
Whenever we figure out our endianness story, we can strip out this size_t and rely on the underlying data being wellformed (saves 8 bytes off a struct, which isn't nothing).

Then we can read the size out of the first 4 bytes of the underlying data when we need to make a view